### PR TITLE
Add an example of non-trivial State usage

### LINKED ISSUE #188: Is it possible to share something like Crossbeam's MsQueue between rocket routes?
Hi, I'm trying to create a production app that accepts data from various endpoints, I parse the requests into some data that I then want to forward to a thread that will commit the data to disk or an external source. I was hoping to use `crossbeam::MsQueue` for the purpose (http://aturon.github.io/crossbeam-doc/crossbeam/sync/struct.MsQueue.html)

I'm not really experienced with concurrency and its limitations with rust (and rocket by implication), but I seem to have been able to get the `MsQueue` shared via rocket's new `State`, but I can't get my consuming thread to work.

Is this possible? Am I on the wrong track? Any recommendations for this kind of scenario?

Thanks!

### PR DESCRIPTION
**Pull request by stephanbuys**
State gives us an opportunity to ship data received asynchronously to a central 'handling' thread, this example may affect performance as crossbeam::sync::MsQueue might introduce a lock (uncertain of this), but the utility of doing this, or similar patterns speaks for itself.

Resolves #188

### CODE REVIEW

File: examples/state_nontrivial/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "state_nontrivial"
+version = "0.1.0"
+authors = ["Stephan Buys <stephan.buys@panoptix.co.za>"]
```
Comment (SergioBenitez) [inline examples/state_nontrivial/Cargo.toml]: No authors in examples, and version should be `0.0.1`. This should be in the workspace via `workspace = "../.."`.

File: examples/state_nontrivial/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+[dependencies]
+crossbeam = "*"
+rocket = { path = "../../lib" }
+rocket_codegen = { path = "../../codegen" }
+
+[dev-dependencies]
+rocket = { path = "../../lib", features = ["testing"] }
+
```
Comment (SergioBenitez) [inline examples/state_nontrivial/Cargo.toml]: Kill this empty line.

File: Cargo.toml [hunk 3]
```diff
@@ -29,6 +29,7 @@ members = [
   "examples/raw_upload",
   "examples/pastebin",
   "examples/state",
+  "examples/state_nontrivial",
```
Comment (SergioBenitez) [inline Cargo.toml]: Lets call this "managed_queue" or something similar.

File: examples/state_nontrivial/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
+#![feature(plugin, custom_derive)]
+#![plugin(rocket_codegen)]
+
+extern crate crossbeam;
+extern crate rocket;
+
+use crossbeam::scope;
+use crossbeam::sync::MsQueue;
+use rocket::State;
+use std::sync::Arc;
```
Comment (SergioBenitez) [inline examples/state_nontrivial/src/main.rs]: Move `std` imports before extern imports, and add a new line between them.

File: examples/state_nontrivial/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    queue.0.push(event);
+    "got it"
+}
+
+// Use with: curl http://<rocket ip>:8000/test?foo=bar
+
+fn main() {
+    let q: Arc<MsQueue<Event>> = Arc::new(MsQueue::new());
+
+    scope(|scope| {
```
Comment (SergioBenitez) [inline examples/state_nontrivial/src/main.rs]: Do we need `scope`? Can we just use regular threads?

File: examples/state_nontrivial/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        scope.spawn(|| {
+            loop {
+                match q.try_pop() {
```
Comment (SergioBenitez) [inline examples/state_nontrivial/src/main.rs]: Why not just `pop`?  `let element = q.pop()`?

File: examples/state_nontrivial/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                    //This is a synchronous stream of received events
+                    Some(e) => println!("Do something useful: {:?}", e),
+                    _ => {
+                        thread::sleep(Duration::from_millis(THREAD_SLEEP))
+                    }
+                }
+            }
+        });
+
+        scope.spawn(|| {
```
Comment (SergioBenitez) [inline examples/state_nontrivial/src/main.rs]: This can be in the main thread, right?

File: examples/state_nontrivial/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    use rocket::http::Method::*;
+    use crossbeam::sync::MsQueue;
+    use std::sync::Arc;
+    use super::LogChannel;
+    use super::Event;
+
+    #[test]
+    fn test_get() {
+        let q: Arc<MsQueue<Event>> = Arc::new(MsQueue::new());
+        let rocket = rocket::ignite().manage(LogChannel(q.clone())).mount("/", routes![super::test]);
```
Comment (SergioBenitez) [inline examples/state_nontrivial/src/main.rs]: Put this in a `rocket()` function in `main.rs` and call `rocket()` from `main.rs` and here.

File: examples/state_nontrivial/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    use rocket::testing::MockRequest;
+    use rocket::http::Status;
+    use rocket::http::Method::*;
+    use crossbeam::sync::MsQueue;
+    use std::sync::Arc;
+    use super::LogChannel;
+    use super::Event;
+
+    #[test]
+    fn test_get() {
```
Comment (SergioBenitez) [inline examples/state_nontrivial/src/main.rs]: What is this testing, exactly? Might be better to have a `put` and `get` route, where `put` does a `push` and `get` does a `pop`. Then we can actually test that the queue is holding things by doing multiple puts and multiple gets.

**Comment by SergioBenitez** [comment]
Waiting on changes here.

**Comment by stephanbuys** [comment]
Apologies, bit snowed under. Will do ASAP

**Comment by SergioBenitez** [comment]
No worries! I know how it goes. :)

**Comment by SergioBenitez** [comment]
@stephanbuys Any update on this?

**Comment by stephanbuys** [comment]
@SergioBenitez apologies for the delay, I think I got all the changes you requested?

**Comment by SergioBenitez** [comment]
@stephanbuys Testing has failed.

**Comment by SergioBenitez** [comment]
Also, do we actually need the `Arc` here?

**Comment by stephanbuys** [comment]
@SergioBenitez - fixed tests and cleaned up the Arc (it was necessary for the initial code).

File: examples/managed_queue/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "managed_queue"
+version = "0.0.1"
```
Comment (SergioBenitez) [inline examples/managed_queue/Cargo.toml]: Can you set this to 0.0.0 to make sure it doesn't accidentally get published to `crates.io`?

**Comment by SergioBenitez** [comment]
Went ahead and made the change myself (and a few other tidy changes). Merged in b4586f62ee9dfea13ff575bf002c8f1f097a9423. Thanks so much! :D

**Comment by stephanbuys** [comment]
@SergioBenitez My pleasure, and thanks for the guidance.

**Pull request closed without merging**